### PR TITLE
Members only authentication project

### PR DESCRIPTION
**Pull request by alfredoC10**
The goal of this project was to understand the foundations of authentication in Ruby on Rails by creating different form layouts that ask the user for different data, so they can have their own account. We set up the user controllers, models and views with the help of the devise gem and render each using bootstrap.

Built with:
- Ruby on Rails
- VS Code
- Devise gem
- Simple forms gem
- Bootstrap
- Rubocop

### CODE REVIEW

**Review by LamiaSristy — CHANGES_REQUESTED** [review]
# Status: Change Required. ♻️

Hello @alfredoC10  and @Wusinho. Well done guys. 

## The highlights: 

1. Correct `git-flow`.💯
2.  Nice and descriptive `README` file. 👍
3.  The Repository is professional with PR message, title, and commit messages.💯
4.  Linters is correct.  👍 👍
5. You have used the `devise` gem for authentication correctly. 👏 👏 👏
6.  Nice job with the message edit and destroy part. ⭐ 

 ## What to do next:
**You need to implement the works below to get the project approved!🛠️**
Please check my points here: 👇

**[REQUIRED]**
- [x] A guest user cannot see the create message form, they only can see the posts. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/guest_can_see_only_posts_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)
screencapture-localhost-3000-2021-02-25-20_40_50

- [x] In `MessagesController`, use `before_action` to restrict access to the `:new` and `:create` methods. `before_action :signed_in_only!, only: [:new, :create]` In your case add also edit and destroy.
- [ ] The `new` method in `MessagesController` it initializes an empty post `@post = Post.new` [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_new_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)
- [x] Remove all the commented lines in `MessagesController`.
- [x] In `routes.rb` file, It creates the application routes using `resources :posts, only: [:new, :create, :index]` [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)
In your case it should be `resources :messages, only: [:new, :create, :index]`

- [ ] No logics should be in views. Use `MessagesHelper` for the if logics.
##

Please make the changes and ask for a re-review.
Feel free to reach me on slack @lamiasristy 🙋

Keep up the spirit always. Happy Coding 💻

File: app/controllers/messages_controller.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  def index
+    @messages = Message.all.order('created_at DESC')
+    @message = Message.new
+  end
+
+  def show; end
+
+  def new
+    @message = current_user.messages.build
+  end
```
Comment (voscarmv) [inline app/controllers/messages_controller.rb]: As the previous TSE mentioned:

>  The new method in MessagesController it initializes an empty post @post = Post.new

File: app/views/messages/_feed.html.erb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  </div>
+  <div class="card-body">
+    <p class="card-text"><%= message.body %></p>
+     <% if user_signed_in? and current_user.id == message.user_id %>
+
+      
+      <%= link_to 'Edit', edit_message_path(message) %>
+      <%= link_to 'Destroy', message, method: :delete, data: { confirm: 'Are you sure?' } %>
+   
+   <% end  %>
```
Comment (voscarmv) [inline app/views/messages/_feed.html.erb]: As the previous TSE mentioned:

>  No logics should be in views. Use MessagesHelper for the if logics.

**Review by voscarmv — CHANGES_REQUESTED** [review]
Impressive work. 💻 👀 ✨ 

- I see you've implemented most of the changes required by the previous TSE. Good work!

## However, there are some changes required 👀 🧰 🔨 🔧 

- Let's take a step back and finish applying the required changes. I've added some code comments to help you find what the previous TSE meant by the unfinished changes. Please fix them!

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @Wusinho  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] You have some logic in views: View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)
  - [ ] There shoud be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright _=> To make sure that I clarify what you need to change on this reqirement: Ensure that there is no if conditions, and no loops in your view (`.html.erb`) files_
  
  => Currently you have logics in the following view files: `app/views/messages/_feed.html.erb` and `app/views/messages/_form.html.erb`
  
#### Resources to help you learn about this:

- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code/)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Comment by Wusinho** [comment]
> # Status: Changes Required  
> ## Hello @Wusinho   
>  This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
>  I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
> 
> ####  However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below   
> ### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding
> # [Required Changes] 
> * [ ]  You have some logic in views: View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)
>   
>   * [ ]  There shoud be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright _=> To make sure that I clarify what you need to change on this reqirement: Ensure that there is no if conditions, and no loops in your view (`.html.erb`) files_
>   
>   => Currently you have logics in the following view files: `app/views/messages/_feed.html.erb` and `app/views/messages/_form.html.erb`
> 
> #### Resources to help you learn about this:
> * [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code/)
> * [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
> 
> _If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._
> 
> [GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)
> 
> _Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

Please I would like you to send me a link because I don't see how I can do what you are asking. I have seen a lot of repos and in all of them the simple_form loop appears on the views. That includes your repo as well.

**Comment by Wusinho** [comment]
> # Status: Changes Required  
> ## Hello @Wusinho   
>  This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
>  I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
> 
> ####  However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below   
> ### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding
> # [Required Changes] 
> * [ ]  You have some logic in views: View-related logic can be encapsulated in View helpers (app/helpers/*_helper.rb)
>   
>   * [ ]  There shoud be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright _=> To make sure that I clarify what you need to change on this reqirement: Ensure that there is no if conditions, and no loops in your view (`.html.erb`) files_
>   
>   => Currently you have logics in the following view files: `app/views/messages/_feed.html.erb` and `app/views/messages/_form.html.erb`
> 
> #### Resources to help you learn about this:
> * [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code/)
> * [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
> 
> _If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._
> 
> [GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)
> 
> _Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

@descholar-ceo 
I understand what you are asking but I can't find how to do it, If you could provide me with some information about it, I'd appreciate it. I have also searched for an answer in other projects' repos and all of them have a simple_form loop by default, that includes your repo as well. I appreciate it if you could provide us with extra help to accomplish the objectives for this project.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @alfredoC10 and @Wusinho, You guys are doing perfectly well, your progress is commendable 👍 Well done!

> Good job on eliminating all the logics from your view 👏 

Let's make a few changes, and we will be done.

- [x] Add validations to your message model, I should not be able to create an empty message
- [x] Do not forget to add `yarn` to your getting started instructions, perfect after the `bundle install`. I couldn't run the project without running `yarn`

I am sorry for bringing this up, but it's good you get familiar with these little concepts and practices in other to avoid them in the future.

Make these changes and you'll be on your way to the next project.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla or reply to this thread), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by voscarmv — APPROVED** [review]
# Project status: Approved ✨ 😎👍 💯 ✔️
- Congratulations, all your hard work has paid off!
- Great work applying all the necessary changes to make your project perfect.
### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**